src/diverse_seq/cluster.py

```diff
@@ -0,0 +1,618 @@
+            distances_chunk[i, j] = distance
+    return start_idx, distances_chunk
+
+
+def mash_sketch(
+    record: DataMember,
+    k: int,
+    sketch_size: int,
+    num_states: int,
+    *,
+    canonical: bool,
+) -> BottomSketch:
+    """Find the mash sketch for a sequence record.
+
+    Parameters
+    ----------
+    record : DataMember
+        The sequence record to find the sketch for.
+    k : int
+        kmer size.
+    sketch_size : int
+        Size of the sketch.
+    num_states : int
+        Number of possible states (e.g. GCAT gives 4 for DNA).
+    canonical : bool
+        Whether to use the canonical representation of kmers.
+
+    Returns
+    -------
+    BottomSketch
+        The bottom sketch for the given sequence record.
+    """
+    seq = record.read()
+    kmer_hashes = {
+        hash_kmer(kmer, canonical=canonical) for kmer in iter_kmers(seq, k, num_states)
+    }
+    heap = []
+    for kmer_hash in kmer_hashes:
+        if len(heap) < sketch_size:
+            heapq.heappush(heap, -kmer_hash)
+        else:
+            heapq.heappushpop(heap, -kmer_hash)
+    return sorted(-kmer_hash for kmer_hash in heap)
+
+
+def compute_mash_distance(
+    left_sketch: BottomSketch,
+    right_sketch: BottomSketch,
+    k: int,
+    sketch_size: int,
+) -> float:
+    """Compute the mash distance between two sketches.
+
+    Parameters
+    ----------
+    left_sketch : BottomSketch
+        A sketch for comparison.
+    right_sketch : BottomSketch
+        A sketch for comparison.
+    k : int
+        kmer size.
+    sketch_size : int
+        Size of the sketches.
+
+    Returns
+    -------
+    float
+        The mash distance between two sketches.
+    """
+    # Following the source code implementation
+    intersection_size = 0
+    union_size = 0
+
+    left_index = 0
+    right_index = 0
+    while (
+        union_size < sketch_size
+        and left_index < len(left_sketch)
+        and right_index < len(right_sketch)
+    ):
+        left, right = left_sketch[left_index], right_sketch[right_index]
+        if left < right:
+            left_index += 1
+        elif right < left:
+            right_index += 1
+        else:
+            left_index += 1
+            right_index += 1
+            intersection_size += 1
+        union_size += 1
+
+    if union_size < sketch_size:
+        if left_index < len(left_sketch):
+            union_size += len(left_sketch) - left_index
+        if right_index < len(right_sketch):
+            union_size += len(right_sketch) - right_index
+        union_size = min(union_size, sketch_size)
+
+    jaccard = intersection_size / union_size
+    if intersection_size == union_size:
+        return 0.0
+    if intersection_size == 0:
+        return 1.0
+    distance = -math.log(2 * jaccard / (1.0 + jaccard)) / k
+    if distance > 1:
+        distance = 1.0
+    return distance
+
+
+def iter_kmers(
```

Comment:
Does this really benefit from being a generator? Also, add a comment about moving this functionality to cogent3.

Reply:
Probably not too much compared to computations elsewhere in all honesty, if you would prefer it to just return a list of the kmers I can do that instead.

Reply:
I suggest have it return a list or a numpy array and rename function to `get_kmers()`